compiler/vectorized_test.go

```diff
@@ -0,0 +1,149 @@
+	obj := make(map[string]values.Value)
+	for k, v := range mp {
+		switch s := v.(type) {
+		case []interface{}:
+			obj[k] = values.NewVectorFromElements(mem, s...)
+		case map[string]interface{}:
+			obj[k] = vectorizedObjectFromMap(v.(map[string]interface{}), mem)
+		default:
+			panic("bad input to vectorizedObjectFromMap")
+		}
+	}
+	return values.NewObjectWithValues(obj)
+}
+
+// Check that:
+//     1. Vectorized inputs yield vectorized outputs when compiled and evaluated
+//     2. The number of bytes allocated is 0 once evaluation is complete
+//     3. Only certain function expressions are vectorized when invoking the
+//        analyzer from go code. The criteria for supported expressions may
+//        change in the future, but right now we only support trivial identity
+//        functions (i.e., those in the form of `(r) => ({a: r.a})`, or something
+//        similar)
+func TestVectorizedFns(t *testing.T) {
+	testCases := []struct {
+		name         string
+		fn           string
+		vectorizable bool
+		allocated    int64
+		maxAllocated int64
+		inType       semantic.MonoType
+		input        map[string]interface{}
+		want         map[string]interface{}
+		skipComp     bool
+	}{
+		{
+			name:         "field access",
+			fn:           `(r) => ({c: r.a, d: r.b})`,
+			vectorizable: true,
+			inType: semantic.NewObjectType([]semantic.PropertyType{
+				{Key: []byte("r"), Value: semantic.NewObjectType([]semantic.PropertyType{
+					{Key: []byte("a"), Value: semantic.NewVectorType(semantic.BasicInt)},
+					{Key: []byte("b"), Value: semantic.NewVectorType(semantic.BasicInt)},
+				})},
+			}),
+			input: map[string]interface{}{
+				"r": map[string]interface{}{
+					"a": []interface{}{int64(1)},
+					"b": []interface{}{int64(2)},
+				},
+			},
+			want: map[string]interface{}{
+				"c": []interface{}{int64(1)},
+				"d": []interface{}{int64(2)},
+			},
+		},
+		{
+			name:         "extend record",
+			fn:           `(r) => ({r with b: r.a})`,
+			vectorizable: true,
+			inType: semantic.NewObjectType([]semantic.PropertyType{
+				{Key: []byte("r"), Value: semantic.NewObjectType([]semantic.PropertyType{
+					{Key: []byte("a"), Value: semantic.NewVectorType(semantic.BasicFloat)},
+				})},
+			}),
+			input: map[string]interface{}{
+				"r": map[string]interface{}{
+					"a": []interface{}{1.2},
+				},
+			},
+			want: map[string]interface{}{
+				"a": []interface{}{1.2},
+				"b": []interface{}{1.2},
+			},
+		},
+		{
+			name:         "no binary expressions",
+			fn:           `(r) => ({c: r.a + r.b})`,
+			vectorizable: false,
+			skipComp:     true,
+		},
+		{
+			name:         "no literals",
+			fn:           `(r) => ({r with c: "count"})`,
+			vectorizable: false,
+			skipComp:     true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			mem := &memory.Allocator{}
+			checked := arrow.NewCheckedAllocator(mem)
```

Comment:
This won't work quite the way you want it to, I think you need to do this:
```go
checked := arrow.NewCheckedAllocator(memory.DefaultAllocator)
mem := &memory.Allocator{Allocator: checked}
```
As it's written now, the checked allocator won't be used in the call to `AssertSize` below.

I would also expect that you need to call `Release` on the input and then also on the output record for these test cases to pass.